File: acwing/算法基础课/第五讲/896.最长上升子序列II/896.最长上升子序列II.cpp
#include <iostream>
#include <cstring>
#include <algorithm>

using namespace std;

const int N = 100000 + 10;

int f[N], a[N];

int main()
{
    int n;
    cin >> n;
    for (int i = 1; i <= n; i++)
        cin >> a[i];

    int len = 0;
    for (int i = 1; i <= n; i++)
    {
        int l = 0;
        int r = len;

        while (l < r)
        {
            int mid = (l + r + 1) >> 1;
            if (f[mid] < a[i])
            {
                l = mid;
            }
            else
            {
                r = mid - 1;
            }
        }
        len = max(len, r + 1);
        f[r + 1] = a[i];
    }
    printf("%d\n", len);
    return 0;
}

File: acwing/算法基础课/第五讲/896.最长上升子序列II/896.最长上升子序列II.py
N = int(1e5) + 10

f = [int(-2e9)] * N

n = int(input())

arr = list(map(int,input().split()))

LEN = 0
for i in range(n):
    l = 0
    r = LEN
    while l < r:
        mid = (l + r + 1) // 2
        if f[mid] < arr[i]:
            l = mid
        else:
            r = mid - 1
    LEN = max(LEN,r + 1)
    f[r + 1] = arr[i]

print(LEN)